kotlin-grpc/src/main/kotlin/com/linecorp/armeria/server/grpc/CoroutineServerInterceptor.kt

```diff
@@ -0,0 +1,77 @@
+package com.linecorp.armeria.server.grpc/*
+ * Copyright 2023 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+import com.linecorp.armeria.internal.common.kotlin.ArmeriaRequestCoroutineContext
+import io.grpc.Metadata
+import io.grpc.ServerCall
+import io.grpc.ServerCallHandler
+import kotlinx.coroutines.DelicateCoroutinesApi
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.asCoroutineDispatcher
+import kotlinx.coroutines.future.future
+import java.util.concurrent.CompletableFuture
+
+/**
+ * A [ServerInterceptor][io.grpc.ServerInterceptor] that is able to asynchronously execute the interceptor without blocking the
+ * caller thread.
+ * For example:
+ * ```kotlin
+ * class AuthInterceptor : CoroutineServerInterceptor {
+ *     override suspend fun <ReqT, RespT> suspendedInterceptCall(
+ *             call: ServerCall<ReqT, RespT>,
+ *             headers: Metadata,
+ *             next: ServerCallHandler<ReqT, RespT>
+ *     ): ServerCall.Listener<ReqT> {
+ *         val result = authorizer.authorize(ServiceRequestContext.current(), headers).await()
+ *         if (result) {
+ *             return next.startCall(call, headers)
+ *         } else {
+ *             throw AnticipatedException("Invalid access")
+ *         }
+ *     }
+ * }
+ * ```
+ */
+interface CoroutineServerInterceptor : AsyncServerInterceptor {
```

Comment:
I figured out that `fun` modifier is necessary to implement this API with a lambda expression.
https://kotlinlang.org/docs/fun-interfaces.html#sam-conversions
```suggestion
fun interface CoroutineServerInterceptor : AsyncServerInterceptor {
```

Reply:
<img width="534" alt="image" src="https://user-images.githubusercontent.com/37795866/223632430-64d9da08-6107-46d6-854b-2fd607a58d34.png">

IDE said `Single abstract member cannot declare type parameters` 😭

Reply:
Please ignore this comment. I realized that SAM can't be used with generic type parameters.